cpp/src/arrow/util/atomic_shared_ptr.h

```diff
@@ -89,5 +89,25 @@ inline void atomic_store(
   *p = r;
 }
 
+template <class T>
+inline bool atomic_compare_exchange_strong(
```

Comment:
templates are implicitly `inline`
```suggestion
bool atomic_compare_exchange_strong(
```

Reply:
All other functions are explicitly marked inline in this file, but I can make the change.